### PR TITLE
[JENKINS-21746] expose slave agent host name

### PR DESCRIPTION
**Pull request by ndeloof**
[JENKINS-21746](https://issues.jenkins-ci.org/browse/JENKINS-21746)

default implementation do return wildcard address (shame that we can't use java.net.InetAddress#anyLocalAddress directly)
LauncherDecorator can overrite this method to return alternate hostname/ip to address network infrastructure constraints

### CODE REVIEW

**Comment by ndeloof** [comment]
@reviewbybees

File: core/src/main/java/hudson/Launcher.java [hunk 1]
```diff
@@ -142,6 +143,20 @@ public Computer getComputer() {
...
+     * Gets the IP address of the host running the jenkins slave agent which do control this Launcher
+     * <p>
+     * In most cases, slave agent do run on localhost, but within a containerized architecture, the
+     * slave agent hostname may be a distinct one. A {@link LauncherDecorator} responsible to setup
+     * such an architecture could then decorate this method to return the adequate slave agent host address.
+     * <p>
+     * Default implementation do return wildcard address (0.0.0.0 or ::0)
+     */
+    public String getSlaveAgentHostAddress() throws IOException {
+        return new InetSocketAddress(0).getAddress().getHostAddress();
```
Comment (jglick) [inline core/src/main/java/hudson/Launcher.java]: Surely this should be wrapped in a `Callable` so that it runs on the slave rather than on master? (It seems to produce `0.0.0.0` when I try it on my Linux machine, but if that is guaranteed then we should just hardcode that string to begin with.)

File: core/src/main/java/hudson/Launcher.java [hunk 2]
```diff
@@ -142,6 +143,20 @@ public Computer getComputer() {
     }
 
     /**
+     * Gets the IP address of the host running the jenkins slave agent which do control this Launcher
+     * <p>
+     * In most cases, slave agent do run on localhost, but within a containerized architecture, the
+     * slave agent hostname may be a distinct one. A {@link LauncherDecorator} responsible to setup
+     * such an architecture could then decorate this method to return the adequate slave agent host address.
```
Comment (jglick) [inline core/src/main/java/hudson/Launcher.java]: Yet existing `Launcher.decorateBy*` methods have not been updated to delegate this method. They should be made to use `DecoratedLauncher`, which must then override this.

File: core/src/main/java/hudson/Launcher.java [hunk 3]
```diff
@@ -142,6 +143,20 @@ public Computer getComputer() {
     }
 
     /**
+     * Gets the IP address of the host running the jenkins slave agent which do control this Launcher
```
Comment (jglick) [inline core/src/main/java/hudson/Launcher.java]: Javadoc summary sentences must be terminated with a period (`.`).

**Comment by jglick** [comment]
`CommandLauncherTest.commandFails` failure, probably unrelated.

File: core/src/main/java/hudson/Launcher.java [hunk 4]
```diff
@@ -142,6 +143,20 @@ public Computer getComputer() {
...
+     * <p>
+     * In most cases, slave agent do run on localhost, but within a containerized architecture, the
+     * slave agent hostname may be a distinct one. A {@link LauncherDecorator} responsible to setup
+     * such an architecture could then decorate this method to return the adequate slave agent host address.
+     * <p>
+     * Default implementation do return wildcard address (0.0.0.0 or ::0)
+     */
```
Comment (jglick) [inline core/src/main/java/hudson/Launcher.java]: `@since TODO`

**Comment by jglick** [comment]
Please avoid rebasing inside a PR as it makes it hard to follow the history of changes vs. reviews. Someone merging can always choose to collapse commits if they prefer.

**Comment by ndeloof** [comment]
ok sorry for rebasing

**Comment by jglick** [comment]
The original issue discussed use of `InetAddress.getLocalHost`, whereas this implementation only offers an address usable from the same machine as the slave agent. If that is an intentional limitation, it should be documented as such.

Also the issue discussed port forwarding, which might in general be necessary for usage from containers. That is not addressed here.

**Comment by jglick** [comment]
:bug: on general principle: this is an API addition without corresponding usage, and it is certainly not obvious to me what exactly it is for or that it would even work. I would expect to see `work-in-progress` PRs in plugins demonstrating its usage end-to-end to solve a problem that cannot be done in any other (reasonable) way.

For example, if you are trying to integrate `docker-custom-build-environment` with `maven-plugin`, so that the `mvn` process inside a container can make a connection to a slave agent on the host, I would expect to see PRs in both of those plugins using a `SNAPSHOT` version of core: one overriding this method to decorate the launcher with the Docker host IP, the other calling this method to find an endpoint for the Maven remoting channel. And of course a functional and/or integration test actually using these plugins together demonstrating that it all works.

**Comment by ndeloof** [comment]
@jglick usage PR from maven-plugin will require https://github.com/jenkinsci/maven-plugin/pull/51 first, as my plan is to include a `TcpSocketHostLocator` to rely on `getSlaveAgentHostAddress`, as a jenkins-core fix to this issue, but also offer an extension point so this can be fixed _today_ without waiting for a future LTS.

If you don't only look at JENKINS-21746 but also https://github.com/jenkinsci/maven-interceptors/pull/4 you will understand this is unrelated to `InetAddress.getLocalHost` but as a generic need to determine slave agent host address, which by default is considered to be "host wildcard interface" (considering maven-plugin, which is the only plugin I know to rely on this). MAVEN_REMOTE_USEINET has been introduced for a comparable need to restrict this to `InetAddress.getLocalHost` to address infrastructure constraints.

**Comment by jglick** [comment]
OK, will just wait for the use cases to be fully prepared, including end-to-end tests.

> also offer an extension point so this can be fixed today without waiting for a future LTS

So there should be a follow-up PR in `maven-plugin` (based on the first) which use a core `SNAPSHOT` and demonstrates how the proposed core change can simplify that plugin.

**Comment by ndeloof** [comment]
proposed core change won't simplify maven-plugin, but will move customization into DecoratedLauncher, as you suggested on https://github.com/jenkinsci/maven-interceptors/pull/4#issuecomment-34653389
The proposed environment variable solution was fine for me, but I try to do it the way you suggested

**Comment by jglick** [comment]
The question is whether the issue can be handled entirely within `maven-plugin` (and its libraries). If so, no core API is needed. A core API would be valuable if that can allow other plugins that work with Docker etc. to avoid an (optional) dependency on an extension point defined in `maven-plugin`.

**Comment by ndeloof** [comment]
This issue can he handled entirely within maven-plugin if you consider some code to set env variable only used by maven plugin to be a reasonable dependency, vs a direct dependency on plugin extension points. I you consider a better option to avoid cross plugin dependencies, then https://github.com/jenkinsci/maven-interceptors/pull/4 is a better choice. 
https://github.com/jenkinsci/maven-interceptors/pull/8 is just a variation, moving the host IP lookup login in maven plugin to avoid code duplication in maven-interceptors (which already is more or less 3 time the same code)

**Comment by jglick** [comment]
> if you consider some code to set env variable only used by maven plugin to be a reasonable dependency

I do not consider that a viable long-term option; it is a backdoor API.

File: core/src/main/java/hudson/Launcher.java [hunk 5]
```diff
@@ -25,11 +25,11 @@
 
 import hudson.Proc.LocalProc;
 import hudson.model.Computer;
+import hudson.remoting.Callable;
 import hudson.util.QuotedStringTokenizer;
 import jenkins.model.Jenkins;
 import hudson.model.TaskListener;
 import hudson.model.Node;
-import hudson.remoting.Callable;
```
Comment (KostyaSha) [inline core/src/main/java/hudson/Launcher.java]: why order of imports changed? Existing looks alphabetically right

**Comment by jglick** [comment]
Looks like you are abandoning a cross-plugin implementation of this? Should JENKINS-21746 be closed?
